---
layout: post
title: 前端进阶知识（88）———— ES6 Generator异步应用
date: 2017-09-19
tags: [ES6]
---

上一篇文章说过，Generator是用来进行异步编程的，那么接下来就介绍它是怎样进行异步操作的。

## Base concept 

传统方法异步编程大概有以下几种：

- 回调函数
- 事件监听 
- Promise对象
- 发布订阅模式（这个具体大家可以自行百度，这里不展开）

接下来介绍Generator实现的异步编程

### 协程

"协程"（coroutine），意思是多个线程互相协作，完成异步任务。

协程有点像函数，又有点像线程。它的运行流程大致如下。

    第一步，协程A开始执行。
    第二步，协程A执行到一半，进入暂停，执行权转移到协程B。
    第三步，（一段时间后）协程B交还执行权。
    第四步，协程A恢复执行。

上面流程的协程A，就是异步任务，因为它分成两段（或多段）执行。

### 协程的Generator 实现

Generator 函数是协程在 ES6 的实现，最大特点就是可以交出函数的执行权（即暂停执行）。

整个 Generator 函数就是一个封装的异步任务，或者说是异步任务的容器。异步操作需要暂停的地方，都用yield语句注明。

    function * gen(x){
        var y = yield x + 3;
        return y;
    }
    var g = gen(1);
    g.next() //{value: 4,done: false}
    g.next() //{value: undefined,done:true}

上面的方法，调用Generator函数，返回一个遍历器（内部指针对象），每次调用next都会将内部指针移动一次，指向第一个遇到的yield，上面执行到x+2为止。

### Generator 函数的数据交换和错误处理

Generator 函数可以暂停执行和恢复执行，这是它能封装异步任务的根本原因。除此之外，它还有两个特性，使它可以作为异步编程的完整解决方案：函数体内外的数据交换和错误处理机制。

next返回值的value属性，是 Generator 函数向外输出数据；next方法还可以接受参数，向 Generator 函数体内输入数据。

    function* gen(x){
    var y = yield x + 2;
    return y;
    }

    var g = gen(1);
    g.next() // { value: 3, done: false }
    g.next(2) // 把2传入 Generator 函数{ value: 2, done: true }

处理错误代码：

    function  * gen(x){
        try{
            var y = yield x + 2;
        }catch(e){
            console.log(e);
        }
        return y;
    }
    var g = gen(1);
    g.next();
    g.throw("getting wrong")
    //getting wrong

，使用指针对象的throw方法抛出的错误，可以被函数体内的try...catch代码块捕获。这意味着，出错的代码与处理错误的代码，实现了时间和空间上的分离，这对于异步编程很重要。

## 异步任务的封装

    var fetch = require("node-fetch");
    function * gen(){
        var url = "https://baidu.com";
        var result = yield fetch(url);
        console.log(result.bio);
    }
    var g = gen();
    var result = g.next();

    result.value.then(function(data){
    return data.json();
    }).then(function(data){
    g.next(data);
    });

第一段代码，读取一个远程接口，从JSON格式的数据解析信息。第二段代码，执行Generator函数，使用next方法，执行异步任务的第一个阶段。

关于Generator就讲到这里吧，如果大家要深入理解，要实战才行。

